virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1251 @@
+ */
+void print_iface_list() {
+	int i;
+
+	for (i = 0; i < MAX_IFACES; i++) {
+		if (iface_list[i].idx == INIT_IF_IDX) {
+			continue;
+		}
+
+		printf("IFACE %d\n", i);
+		printf("\tidx     = %d\n", iface_list[i].idx);
+		printf("\thw_addr = %s\n", iface_list[i].hw_addr);
+		printf("\tname    = %s\n", iface_list[i].name);
+		printf("\tmtu     = %s\n", iface_list[i].mtu);
+
+		struct ip_addr *ipa = iface_list[i].ip_addrs;
+		if (!ipa) {
+			printf("\n");
+			continue;
+		}
+
+		printf("\tip_addrs\n");
+		printf("\t    |---- ip_addr = %s, family = %d\n",
+				ipa->addr, ipa->family);
+		while (ipa->next) {
+			printf("\t    |---- ip_addr = %s, family = %d\n",
+					ipa->next->addr, ipa->next->family);
+			ipa = ipa->next;
+		} 
+
+		printf("\n");
+	}
+}
+
+/*
+ * Print program usage.
+ */
+void print_usage(void)
+{
+	printf("Usage: %s [options]\n\n", PROGRAM_NAME);
+	printf(" -d, --debug        Enable debug output\n");
+	printf(" -h, --help         Display usage\n");
+	printf(" -n, --netns-path   Network namespace path (required)\n");
+	printf(" -p, --pod-id       Pod ID (required)\n");
+	printf(" -r, --runtime-path Runtime path (required)\n");
+	printf(" -v, --version      Show version\n");
+	printf("\n");
+}
+
+/*
+ * Print version information.
+ */
+void print_version(void)
+{
+	printf("%s v0.1\n", PROGRAM_NAME);
+}
+
+/*
+ * Free internal fields of the watcher_params structure.
+ */
+void free_watcher_params(struct watcher_params *params)
+{
+	if (!params) {
+		return;
+	}
+
+	free(params->netns_path);
+	params->netns_path = NULL;
+
+	free(params->pod_id);
+	params->pod_id = NULL;
+
+	free(params->runtime_path);
+	params->runtime_path = NULL;
+}
+
+/*
+ * Free internal fields of the ip_addr structure.
+ */
+void free_ip_addr(struct ip_addr *ipa) {
+	struct ip_addr *ipa_next, *ipa_curr;
+
+	if (!ipa) {
+		return;
+	}
+
+	free(ipa->addr);
+	ipa->addr = NULL;
+
+	for (ipa_curr = ipa->next; ipa_curr; ipa_curr = ipa_next) {
+		ipa_next = ipa_curr->next;
+
+		free(ipa_curr->addr);
+		ipa_curr->addr = NULL;
+
+		free(ipa_curr);
+		ipa_curr = NULL;
+	}
+}
+
+/*
+ * Free internal fields of the iface structure.
+ */
+void free_iface(struct iface *nif) {
+	int i;
+
+	if (!nif) {
+		return;
+	}
+
+	free(nif->hw_addr);
+	nif->hw_addr = NULL;
+
+	free(nif->name);
+	nif->name = NULL;
+
+	free(nif->mtu);
+	nif->mtu = NULL;
+
+	if (nif->ip_addrs) {
+		free_ip_addr(nif->ip_addrs);
+	}
+	free(nif->ip_addrs);
+	nif->ip_addrs = NULL;
+}
+
+/*
+ * Free internal fields of the route structure.
+ */
+void free_route(struct route *rt) {
+	free(rt->src);
+	rt->src = NULL;
+
+	free(rt->dst);
+	rt->dst = NULL;
+
+	free(rt->gw);
+	rt->gw = NULL;
+
+	free(rt->dev);
+	rt->dev = NULL;
+}
+
+/*
+ * Modify the global list of interfaces by inserting a new ip_addr structure
+ * into the linked list of the specified interface.
+ */
+void insert_ip_addr(int idx, char addr[], unsigned char family)
+{
+	struct ip_addr *ipa = (struct ip_addr *)
+		malloc(sizeof(struct ip_addr));
```

Comment:
Unless there is a very good reason not to, I'd recommend using `calloc()` always.

Reply:
you're right, done!